src/System.Net.Security/src/System/Net/Security/SNIHelper.cs

```diff
@@ -0,0 +1,251 @@
+            }
+        }
+
+        // 3.1. https://www.ietf.org/rfc/rfc3546.txt
+        // ServerNameList structure:
+        //   - ServerName server_name_list<1..2^16-1>
+        // ServerName structure:
+        //   - NameType name_type (1 byte) => 0x00 is host_name
+        //   - opaque HostName
+        // Per spec:
+        //   If the hostname labels contain only US-ASCII characters, then the
+        //   client MUST ensure that labels are separated only by the byte 0x2E,
+        //   representing the dot character U+002E (requirement 1 in section 3.1
+        //   of [IDNA] notwithstanding). If the server needs to match the HostName
+        //   against names that contain non-US-ASCII characters, it MUST perform
+        //   the conversion operation described in section 4 of [IDNA], treating
+        //   the HostName as a "query string" (i.e. the AllowUnassigned flag MUST
+        //   be set). Note that IDNA allows labels to be separated by any of the
+        //   Unicode characters U+002E, U+3002, U+FF0E, and U+FF61, therefore
+        //   servers MUST accept any of these characters as a label separator.  If
+        //   the server only needs to match the HostName against names containing
+        //   exclusively ASCII characters, it MUST compare ASCII names case-
+        //   insensitively.
+        private static string GetSniFromServerNameList(ReadOnlySpan<byte> serverNameListExtension, out ReadOnlySpan<byte> remainingBytes)
+        {
+            if (serverNameListExtension.Length < 2)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            ushort serverNameListLength = ReadUint16(serverNameListExtension);
+            ReadOnlySpan<byte> serverNameList = serverNameListExtension.Slice(2);
+
+            if (serverNameListLength > serverNameList.Length)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            remainingBytes = serverNameList.Slice(serverNameListLength);
+            ReadOnlySpan<byte> serverName = serverNameList.Slice(0, serverNameListLength);
+
+            if (serverName.Length < 3)
+            {
+                return null;
+            }
+
+            // -1 for hostNameType
+            int hostNameStructLength = ReadUint16(serverName) - 1;
+            byte hostNameType = serverName[2];
+            ReadOnlySpan<byte> hostNameStruct = serverName.Slice(3);
+
+            if (hostNameStructLength != hostNameStruct.Length || hostNameType != 0x00)
+            {
+                return null;
+            }
+
+            ushort hostNameLength = ReadUint16(hostNameStruct);
+            ReadOnlySpan<byte> hostName = hostNameStruct.Slice(2);
+
+            return Encoding.UTF8.GetString(hostName);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ushort ReadUint16(ReadOnlySpan<byte> bytes)
```

Comment:
actually we could leave it as int since we have to double convert ushort otherwise (here and for comparison)